segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/MemoryStateUpdater.java

```diff
@@ -133,32 +166,71 @@ boolean process(Operation operation) throws DataCorruptionException {
             throw new DataCorruptionException("About to have added a Log Operation to InMemoryOperationLog that was out of order.");
         }
 
-        return true;
+        return segmentId;
+    }
+
+    private void logException(Throwable ex) {
+        log.error("Unhandled exception.", ex);
     }
 
     /**
-     * Flushes recently appended items, if needed.
-     * For example, it may trigger Future Reads on the ReadIndex, if the readIndex supports that.
+     * Commits pending CommitGroups.
      */
-    void flush() {
-        HashSet<Long> elements;
-        synchronized (this.readIndex) {
-            elements = this.recentStreamSegmentIds;
-            this.recentStreamSegmentIds = new HashSet<>();
+    private void commitPending() {
+        HashSet<Long> segmentIds = new HashSet<>();
+
+        while (true) {
+            // Process all CommitGroups up to, and including the one identified by commitUpTo.
+            CommitGroup toProcess;
+            synchronized (this.pendingCommits) {
+                if (!this.pendingCommits.isEmpty() && this.pendingCommits.peekFirst().id <= this.commitUpTo) {
+                    toProcess = this.pendingCommits.removeFirst();
+                } else {
+                    break;
+                }
+            }
+
+            // Process all Operations within the CommitGroup in sequence.
+            for (CompletableOperation operation : toProcess.operations) {
+                try {
+                    long segmentId = process(operation.getOperation());
+                    if (segmentId >= 0) {
+                        segmentIds.add(segmentId);
+                    }
+                } catch (Throwable ex) {
+                    toProcess.commitErrorHandler.accept(ex);
```

Comment:
I can see the current behaviour is the same as the old behavior. But are you convinced in the correctness of this error handling? It is non-obvious to me.